Benchmark parameters I

arximboldi opened the pull request:
Hi!

I managed to get started with the feature. There is work to be done, but I open the pull request in case you want to do an early review so I can change my approach if needed. Note that this branch depends on the changes in #54 

**Here's the progress:**
- [x] Benchmark writers can declare one or more parameters of any streamable type.
- [x] Benchmark parameters can be set to a value from the command line.
- [x] The benchmark runner can be told from the command line try a whole range (linear or exponential) of range values for one (and only one) parameter.
- [x] Improve the handling of parameter types.
- [x] Testing (I know, one should do that first, but the areas of the code I needed to touch did not have any tests yet so I just hacked my way through.)
- [x] Update the HTML reporter to show a graph with the different means and variances for the test run with different parameters.

**Nice to have that I might not do:**
- [ ] Update documentation mentioning this feature.
- [ ] Do something special for other reporters when trying a parameter range.
- [ ] Improve parameter parsing on the command line so one can escape the separator `:` in string parameters.
- [ ] Add runs through ranges of values over multiple parameters, using support for 3D and/or contour plots from Plotly.
- [ ] Explicitly defining benchmark suites (capturing parameters at suite level for brevity) 

RFC @rmartinho

arximboldi committed:
Define basic API for declaring and using parameters

arximboldi committed:
Implement pasing parameters to benchmarks

The user may declare parameters by doing, at the top level:
```cpp
    NONIUS_PARAM(name, <default_value>);
```

This declares a parameter.  Parameters can be then passed to the `main`
program by passing arguments of the form:
```
    -p <name>:<value>
```

The parameters have to be accessed from the chronometer.  Since
parameters can have any type and are string-addressed, they can only be
finally parsed or unboxed and looked up in the benchmark.  To stop users
from mistakenly adding that cost to their tests, they are forced to
extract the parameters before, from the chronometer.  Of course they can
cheat by capturing the chronometer, but the documentation should guide
users in the right direction.  A good example:
```
NONIUS_BENCHMARK("push_back vector", [](nonius::chronometer meter)
{
    auto n = meter.param("size");
    auto storage = std::vector<std::vector<char>>(meter.runs());
    meter.measure([&](int i) {
        auto& x = storage[i];
        for (auto j = 0; j < n; ++j)
            x.push_back(static_cast<char>(j));
    });
})
```

arximboldi committed:
Do not magically stringify the declared parameter name

The parameter name is now required to be a string already.  There is no
fundamental reason why parameter names should be limited to identifier
names rules.  The fact that `NONIUS_DETAIL_UNIQUE_NAME` already exists
is convenient here.

This changes syntax for the user from:
```cpp
    NONIUS_PARAM(name, 42);
```

To:
```cpp
    NONIUS_PARAM("name", 42)
```

Note the semicolon is not needed anymore either.

arximboldi committed:
Add parsing of parameters that should be tested at different sizes

The syntax is:
```
    -p <name>:<oper>:<init>:<delta>:<steps>
```

This tells that the benchmark should be run with `<steps>` different
values for parameter `<name>`.  Each value is computed as:
```
    next = last <oper> deta
```

Where `<oper>` may be `+` or `*` as for testing with linear or
exponential increments (the later is useful when checking logarithmic
algorithms or subtle non-linearities in the "constant factors" of an
implementation).

arximboldi committed:
Simplify parsing of running parameters configuration

The old code was convoluted, with too many assertions and implicit
assumptions.  Instead of generating all the samples that we want to try
in the argument parsing, we delay this logic to a more appropriate place
later, without loss of information.

The only drawback of the new configuration representation is that it may
be less flexible, as the previous one allowed to easily add passing
explicitly all the various parmeters to try.  We can, though, easily add
this later by adding more options to the configuration.

arximboldi committed:
Run through a parameter range when the option is enabled

arximboldi committed:
Simplify go() function

The functions `go_benchmark` and `go_param` where extracted, since the
function had grown too big to my taste after the parameter running logic
had been introduced.

Also, the main loop has been simplified to avoid the need to `continue`,
which is more structured IMHO.

rmartinho commented:
The build server cannot build this PR because the author is not whitelisted. A meatbag will have to look at it.

**This is an automated message.**

arximboldi committed:
Perform operations on parameters in their declared value

We always store parameters generically as `std::string`.  The user
implicitly suggests a concrete value type with the type of the default
value of the parameter.  We use that type when we need to do arithmetic
on the parameter and also when checking that it parses correctly when
passed from the command line.  We do this by casting the strings back to
the original `T` using the `param_spec` interface.

This might seem overkill.  Another approach could be to type erase the
concrete values once intially parsed from the command line, using a
`boost::any`-like type erasure class.  However that approach is less
flexible, because it requires the user to provide the specific concrete
type when calling `chronometer::param<T>()` and just a mistake in
signedness or whatever could be fatal.  Just parsing everything into and
back from strings is more forgiving on numeric types, which is the most
common case.

This commit adds a new `example7` that where using ranged parameter
would not work correctly previously.

arximboldi committed:
Enable parameters that only support arithmetic partially

In the last commit, we broke parameters of types that do not support `+`
and `*` operators.  Some of these parameters might be useful, for
example `std::string`.

Now, parameters may or may not support these operations.  When they do
not support, the benchmarks will fail when we try to run the parameter
through a range with those operations.

Funny, we can do:
```
   ./bin/examples/example6 -p string:+:hola-mundo:-yeah:10
```

But we can't:
```
   ./bin/examples/example6 -p string:*:hola-mundo:-oh,no!:10
```

arximboldi committed:
Run all benchmarks for a parameter setting in one round

This commit refactors a lot of the code added in this branch. The
original goal was to reverse how `go()` runs through various
parameter values.  Before it did like:

    for each benchmark
        for each parameter
            run-benchmark()

Now it is more like

    for each parameter
        for each benchmark
            run-benchmark()

This new structure should make writing reporters easier, since it better
fits how the data could be presented to the user in meaningful ways --
we'll see more of this in later commits.

A couple of more refactorings has sneaked in this commit:

- `param_map` and `param_registry` etc. are not in `detail::` anymore.
  This is to better fit other parts of the library (e.g. benchmark,
  reporter) that are not in `detail` and arguably at the same level of
  abstraction from the user API POV.

- A couple of functions now have a more "data driven" approach, trying
  to reuse previously computed data more often, as to improve the
  overall performance.

arximboldi committed:
Rename param_map to parameters

rmartinho commented:
Thanks for putting this up so I can review it. I hope I can get some time on sunday. If you want you can rebase on devel now that I merged the other PR.

arximboldi committed:
Add API for writing benchmarks that support reusable fixtures

The old API for benchmarks that have fixtures is like this:
```cpp
NONIUS_BENCHMARK([] (nonius::chronometer meter){
  // setup code
  m.measure([&] {
      // benchmark code
  })
});
```

That API is nice, but since it has to assume that the benchmark code can
always mutate the setup, the setup can not be reused across samples.
This is ok in general, but for benchmarks that have expensive setups
relative to their measured time, time might be wasted uneedesly.

Our alternative syntax is like:
```cpp
NONIUS_BENCHMARK([] (nonius::parameters meter){
  // immutable setup code
  return [=] {
      // benchmark code
  }
});
```

Or even further:
```cpp
NONIUS_BENCHMARK([] (nonius::parameters params){
  // immutable setup code
  return [=] (nonius::chronometer meter) {
      // mutable setup code
      meter.measure([&] {
          // benchmark code
      })
  }
});
```

The idea is that when the benchmark function can take a `parameters`
argument, it is assumed to be a factory that then returns the actual
benchmark function.  This factory function is called *only once* for a
given set of parameters (in `prepare(...)`) and the resulting benchmark
function is kept in the `execution_plan` and reused for collecting all
the samples.

Furthermore, this syntax may be more intuitive for people that need
parameters, but do not need a chronometer.

arximboldi committed:
Prevent infinite loop in run_for_at_least

When trying to figure out how long to run a test to get a meaningful
value, we may fall into an infinite loop when the whole benchmark has
been optimized away.  We now try to detect this case by hard-limiting
the number of iterations we do in the loop.

This partly solves #52.  A complete solution involves properly
deoptimizing the return value of the benchmark function.

arximboldi committed:
Improve output of standard_reporter on error

The exception message would get mixed together with other lines of
messages creating confusion.

arximboldi committed:
Test changes in detail::benchmark_function

arximboldi committed:
Test new parameters API

arximboldi committed:
Test parameter run generation

arximboldi committed:
Fix passing both fixed parameters and a parameter run

Now all the logic of deciding which parameters to execute is contained
in `generate_params`.  We removed the "feature" that `parameters` looks
up default values.  This way the code we remove dependencies on global
state, and make the feature more testable.  `generate_params` now merges
the default values with the fixed values with the values generated for a
parameter run.

arximboldi committed:
Fix parsing multiple instances of the same command line argument

The command line parser would only look at the first instance of an
argument.  This would prevent us from passing multiple parameters to the
benchmark runner. Examples that did not work but now do work:
```
    # Would only set string=foo, discarding size
    example6 -p string:foo -p size:42
```
```
    # Would only run size:42,52,..., discarding string
    example6 -p size:+:42:10:10 -p string:foo
```

arximboldi committed:
Make HTML reporter show summary page

The HTML reporter has been changed to generate multiple plots instead of
just one.  The plots may be choosed with a drop-down menu at the top.
The drop down menu is always focused such that the user may use the
keyboard arrows to quickly navigate the plots.

These are the plots that are generated:

**summary plot:**

  - If the benchmarks are run with only one set of parameters, it is a bar
    graph showing the mean and stddev of each benchmark.

  - If the benchmarks are run with against various parameters, it is a
    line graph showing the evolution of the benchmark results for the
    different parameters.  The graph is shown in logarithmic scale if
    when the parameters are generated using `*`.

**samples plot:**

   - For every set of parameters, a plot is generated depicting every
     sample taken for every benchmark.  This plot is like the one that
     was generated before this change.

arximboldi commented:
Hi @rmartinho!

Finally, I removed the `wip` tag and officially submit this for code review and pull request. Some comments:

- The history contains certain degree of back-and-forth in the design as I was exploring different options and getting to learn better the Nonius architecture. Some commits could be squashed for brevity and clarity, but the current shape may document better the rationale for certain decisions. Since I do not know what kind of history shape you prefer, I submit it "as is". Please tell me if you find it hard to review and I will try my best to improve the history.

- I did not update the documentation page with the feature, because some maintainers prefer to do this themselves as to keep a consistent writing style. Please tell me if you think I should do this myself.

- I did end up refactoring a few parts of the library. I tried my best to do them "in the spirit" of the code of the rest of the library, but I'd be happy to revise the design if my proposal subverts your vision for the future of Nonus. 

- I tried hard to make the feature not affect any code already working with Nonius. The only observable difference is that the HTML reporter now generates a different output, even for benchmarks that do not use parameters. Here is the new HTML output of `example3`:

  https://sinusoid.es/misc/nonius/params-example3.html

  The difference is that now there is a summary page showing a bar graph with the results of the analysis (mean and standard deviation for each benchmark). The old samples graph can be reached by using the drop down menu at the top, or the arrow keys.

  Of course, this new design shines most when parameters are involved. Here is an example comparing pushing back different amount of elements on `vector<char>` or `std::string`, in logarithmic scale (using `--param size:*:1:10:7`). The samples for each run can still be reached in the same way:

    https://sinusoid.es/misc/nonius/params-example7-log.html

arximboldi commented:
Hi @rmartinho!

Finally, I removed the `wip` tag and officially submit this for code review and pull request. Some remarks:

- The history contains certain degree of back-and-forth in the design as I was exploring different options and getting to learn better the Nonius architecture. Some commits could be squashed for brevity and clarity, but the current shape may document better the rationale for certain decisions. Since I do not know what kind of history shape you prefer, I submit it "as is". Please tell me if you find it hard to review and I will try my best to improve the history.

- I did not update the documentation page with the feature, because some maintainers prefer to do this themselves as to keep a consistent writing style. Please tell me if you think I should do this myself.

- I did end up refactoring a few parts of the library. I tried my best to do them "in the spirit" of the code of the rest of the library, but I'd be happy to revise the design if my proposal subverts your vision for the future of Nonus.

- I tried hard to make the feature not affect any code already working with Nonius. The only observable difference is that the HTML reporter now generates a different output, even for benchmarks that do not use parameters. Here is the new HTML output of `example3`:

  https://sinusoid.es/misc/nonius/params-example3.html

  The difference is that now there is a summary page showing a bar graph with the results of the analysis (mean and standard deviation for each benchmark). The old samples graph can be reached by using the drop down menu at the top, or the arrow keys.

  Of course, this new design shines most when parameters are involved. Here is an example comparing pushing back different amount of elements on `vector<char>` or `std::string`, in logarithmic scale (using `--param size:*:1:10:7`). The samples for each run can still be reached in the same way:

  https://sinusoid.es/misc/nonius/params-example7-log.html

So, this is it! It was fun diving into the Nonius code, it is well written and a very valuable tool. Thanks a lot for building it.

arximboldi commented:
Hi @rmartinho!

Finally, I removed the `wip` tag and officially submit this for code review and pull request. Some remarks:
- The history contains certain degree of back-and-forth in the design as I was exploring different options and getting to learn better the Nonius architecture. Some commits could be squashed for brevity and clarity, but the current shape may document better the rationale for certain decisions. Since I do not know what kind of history shape you prefer, I submit it "as is". Please tell me if you find it hard to review and I will try my best to improve the history.
- I did not update the documentation page with the feature, because some maintainers prefer to do this themselves as to keep a consistent writing style. Please tell me if you think I should do this myself.
- I did end up refactoring a few parts of the library. I tried my best to do them "in the spirit" of the code of the rest of the library, but I'd be happy to revise the design if my proposal subverts your vision for the future of Nonus.
- I tried hard to make the feature not affect any code already working with Nonius. The only observable difference is that the HTML reporter now generates a different output, even for benchmarks that do not use parameters. Here is the new HTML output of `example3`:
  
  https://sinusoid.es/misc/nonius/params-example3.html
  
  The difference is that now there is a summary page showing a bar graph with the results of the analysis (mean and standard deviation for each benchmark). The old samples graph can be reached by using the drop down menu at the top, or the arrow keys.
  
  Of course, this new design shines most when parameters are involved. Here is an example comparing pushing back different amount of elements on `vector<char>` or `std::string`, in logarithmic scale (using `--param size:*:1:10:7`). The samples for each run can still be reached in the same way:
  
  https://sinusoid.es/misc/nonius/params-example6-log.html

So, this is it! It was fun diving into the Nonius code, it is well written and a very valuable tool. Thanks a lot for building it.

rmartinho commented:
This looks pretty promising! Thanks a lot! I'll try to give it a thorough review before I go on holiday, but just in case I don't get to finish that on time, don't think I'm ignoring this PR if I don't comment on it for a while.

To address some of your concerns:
- Don't worry about the history. Sometimes I do find myself obsessing a bit about having a clean history, but I don't really think that's a valuable use of my time.
- I want to update the documentation directly myself. It ensures I get an understanding of whatever features people contribute, and it helps me ensure no malicious JS slips through or something like that.
- I really like the output changes!

arximboldi commented:
Thanks! No problem with the timing, take your time. In the meantime I will be using the branch in the project that motivated the change, which might help polishing it.

arximboldi committed:
Fix missing override warning

arximboldi committed:
Fix large numbers would be truncated incorrectly in HTML reporter

rmartinho commented on examples/example7.c++:
```diff
@@ -0,0 +1,79 @@
+NONIUS_PARAM("x", 1.0)
+
+template <typename Fn>
+struct volatilize_fn {
```
I suppose this is here because of #52, right? If it works, I'd say it should be part of what nonius does when running the code and not in the examples. Basically, I think that any such technique that can be applied generically should be built-in.

I have been researching a bit on optimizer control, and I put up some code for that (#57). If you could have a look at it, that'd be great.

rmartinho commented on examples/example6.c++:
```diff
@@ -0,0 +1,35 @@
+NONIUS_BENCHMARK("push_back vector", [](nonius::chronometer meter)
+{
+    auto n = meter.param("size");
+    auto s = meter.param<std::string>("string");
```
While I don't mind the stringly typed parameters too much, someone made me aware of a technique that can use the compiler to ensure type-safety by using type tags instead of strings for the keys. I'll try it out and if it works I'll make that change before merging this.

arximboldi replied:
Lol, I just sent you an email on this topic. I copy here the proposal so other people can see and discuss what's best.

---

The current API forces the user to write the type the type and string name of the parameter on every benchmark.  That feels now too "dynamic", and was also preventing some more elegant implementations of type erasure could be otherwise added without this API.  Here is the change proposal:

**Option 1**:  macro that expands to a type:

``` cpp
NONIUS_PARAM(N, std::size_t{10}) // defines a type N that encapsulates
                                 // the information about the parameter

NONIUS_BENCHMARK("foo", [] (nonius::parameters params) {
    auto n = params.get<N>();    // decltype(n) == std::size_t
})

```

**Option 2**: no macros, global object instead of type

``` cpp
nonius::param<std::size_t> N{"N", 10};

NONIUS_BENCHMARK("foo", [] (nonius::parameters params) {
    auto n = params.get(N);      // decltype(n) == std::size_t
})
```

I was considering that the `N` object could itself provide access to the parameter value, but that would mean that global state is required and benchmarks could not be run in parallel (well, maybe yest with a `thread_local` global, still, global state makes me unconfortable :-).

Note that the two options can be combined.  Also, in either case, I-d like to add higher order functions to make param extraction terser. e.g.

``` cpp
NONIUS_PARAM(N, std::size_t{10})

NONIUS_BENCHMARK("foo", nonius::parametrized<N>([] (auto n) {
    ...
})
```

griwes replied:
I had the first option in mind when suggesting this elsewhere, but I must say that I like the second option too, and I can't say I really considered it before for my own uses of this technique. Thanks for the inspiration!

rmartinho commented:
It might be important for nonius to know which benchmarks parameters are used by each benchmark. As is, all benchmarks are repeated for all parameter values, even if they don't depend on them. This is fine if you group sets of benchmarks each in its own runner but not so nice if you put different ones together. I'm not sure how important such a use case would be, but I see having to split benchmarks across multiple runners as an inconvenience given that the runners have filtering capabilities.

I'm not sure how to have that work, but maybe some kind of benchmark grouping is needed. Just bringing it up for discussion.

arximboldi commented:
@rmartinho with regard to grouping. I also thought of that. The best solution I thought of so far is to add a notion of "suite". Parameters can then be scoped to a suite. Furthermore, this syntax might make it easier to use parameter objects as proxy to query the parameters without runtime cost nor global state. Here is an example of how benchmarks could look like:

``` cpp
NONIUS_SUITE("push_back", [] {
    NONIUS_PARAM(N, std::size_t{10});
    NONIUS_BENCHMARK("vector", [=] {
        for (auto i = 0u; i < N; ++i) 
              ...
    })
    NONIUS_BENCHMARK("string", [=] {
        for (auto i = 0u; i < N; ++i) 
             ...
    })
})
```

I was hesitant to implement that in this branch because it requires even deeper changes to the runner and registration system, possibly breaking compatibility. If you like it I can try implementing it to see how it feels...

ThePhD commented:
If I could throw my two cents in here...

Y E S .

Grouping is a BIG thing for me. I was going to implement it myself when I had time, because I was curious to see how well many of the library I benched performed when you scaled up a certain parameter, and when you compared it across libraries. Replacing the x-axis value of "sample #" for the default graph with instead a kind of "grouping A", "grouping B", and (for the HTML graph bits) doing a scatter-bar there in to represent all the samples of that grouping would be immensely helpful.

(I ended up using matplotlib with python to create a fancier graph to do what I want: this is what it ended up looking like: http://sol2.readthedocs.io/en/latest/benchmarks.html)

rmartinho commented:
Let's try that on a separate branch (though you can branch off of this one if you want; this will definitely be merged). See issue #59.

arximboldi commented:
Cool. I just pushed a little fix on this branch for an old bug. I will move further experimenting with syntax to a new branch.

rmartinho commented on include/nonius/param.h++:
```diff
@@ -0,0 +1,176 @@
+// Nonius - C++ benchmarking tool
+//
+// Written in 2014 by Martinho Fernandes <martinho.fernandes@gmail.com>
```
Ha, I see you copied these from some existing stale ones. Since this is PD this notice is somewhat irrelevant, but if you want to claim credit for this file, you're welcome to do so. Otherwise I'll probably do a sweeping change in the repo to something in the spirit of "by the nonius contributors nonius@rmf.io"

arximboldi replied:
Hehe, yup! 

I think long term it's easiest to put something like _"by the nonius contributors nonius@rmf.io, see AUTHORS file."_ Then the `AUTHORS` file may list maintainers, contributors, etc.
